src/huggingface_hub/hf_file_system.py

```diff
@@ -282,56 +260,149 @@ def ls(
         return out if detail else [o["name"] for o in out]
 
-    def _iter_tree(self, path: str, revision: Optional[str] = None):
-        # TODO: use HfApi.list_files_info instead when it supports "lastCommit" and "expand=True"
-        # See https://github.com/huggingface/moon-landing/issues/5993
+    def _ls_tree(
+        self,
+        path: str,
+        recursive: bool = False,
+        detail: bool = False,
+        refresh: bool = False,
+        revision: Optional[str] = None,
+    ):
         resolved_path = self.resolve_path(path, revision=revision)
-        path = f"{self._api.endpoint}/api/{resolved_path.repo_type}s/{resolved_path.repo_id}/tree/{safe_quote(resolved_path.revision)}/{resolved_path.path_in_repo}".rstrip(
-            "/"
-        )
-        headers = self._api._build_hf_headers()
-        yield from paginate(path, params={"expand": True}, headers=headers)
+        path = resolved_path.unresolve()
+        root_path = HfFileSystemResolvedPath(
+            resolved_path.repo_type, resolved_path.repo_id, resolved_path.revision, ""
+        ).unresolve()
+
+        out = []
+        if path in self.dircache and not refresh:
+            cached_path_infos = self.dircache[path]
+            out.extend(cached_path_infos)
+            dirs_not_in_dircache = []
+            if recursive:
+                # Use BFS to traverse the cache and build the "recursive "output
+                # (The Hub uses a so-called "tree first" strategy for the tree endpoint but we sort the output to follow the spec so the result is (eventually) the same)
+                dirs_to_visit = deque(
+                    [path_info for path_info in cached_path_infos if path_info["type"] == "directory"]
+                )
+                while dirs_to_visit:
+                    dir_info = dirs_to_visit.popleft()
+                    if dir_info["name"] not in self.dircache:
+                        dirs_not_in_dircache.append(dir_info["name"])
+                    else:
+                        cached_path_infos = self.dircache[dir_info["name"]]
+                        out.extend(cached_path_infos)
+                        dirs_to_visit.extend(
+                            [path_info for path_info in cached_path_infos if path_info["type"] == "directory"]
+                        )
+
+            dirs_not_expanded = []
+            if detail:
+                # Check if there are directories with non-expanded entries
+                dirs_not_expanded = [self._parent(o["name"]) for o in out if o["last_commit"] is None]
+
+            if (recursive and dirs_not_in_dircache) or (detail and dirs_not_expanded):
+                # If the dircache is incomplete, find the common path of the missing and non-expanded entries
+                # and extend the output with the result of `_ls_tree(common_path, recursive=True)`
+                common_path = os.path.commonprefix(dirs_not_in_dircache + dirs_not_expanded).rstrip("/")
```

Comment:
According to [Python docs](https://docs.python.org/3/library/os.path.html#os.path.commonprefix), `common_path` is not guaranteed to be a valid path. For example,

```py
>>> import os
>>> os.path.commonprefix(["data/train/", "data/test/"])
"data/t"
```

which can lead to an unexpected result in `self._ls_tree(common_path, ...)` just below.

I saw there is [`os.path.commonpath`](https://docs.python.org/3/library/os.path.html#os.path.commonpath) as well but I'm afraid it will not work on Windows. What we can do is to use [`PurePosixPath`](https://docs.python.org/3/library/pathlib.html#pathlib.PurePosixPath) like this:

```py
from pathlib import PurePosixPath
from typing import List


def unix_commonpath(paths: List[str]) -> str:
    """Return the longest common subpath of a list of Unix paths."""
    common_parents = set.intersection(
        *[set(parent.as_posix() for parent in PurePosixPath(path).parents) for path in paths]
    )
    return max(common_parents, key=lambda p: len(p))
```

```python
>>> unix_commonpath(["data/train/", "data/test/"])
"data"
```

Let me know if you think of a simpler solution. If we go with this one, could you add a few test cases to ensure it works cross-platform?

Reply:
Good catch!

Reply:
There is a simpler fix to this. I added a test to ensure it works.